src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1190,6 +1192,9 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     val self1 = typed(self)(ctx.outer).asInstanceOf[ValDef] // outer context where class members are not visible
     val dummy = localDummy(cls, impl)
     val body1 = typedStats(impl.body, dummy)(inClassContext(self1.symbol))
+
+    typedUsecases(body1.map(_.symbol), self1.symbol)(localContext(cdef, cls).setNewScope)
```

Comment:
@odersky: using a fresh scope here, also got rid of the flags and changed the `lazy val` to a `def`.  So if this is correct - then this PR should be good to go :)